azure-kusto-ingest/azure/kusto/ingest/ingest_client.py

```diff
@@ -108,3 +125,32 @@ def ingest_from_blob(self, blob_descriptor: BlobDescriptor, ingestion_properties
         content = ingestion_blob_info_json
         queue_client = queue_service.get_queue_client(queue=random_queue.object_name, message_encode_policy=TextBase64EncodePolicy())
         queue_client.send_message(content=content)
+
+    def validate_endpoint_service_type(self):
+        if not hasattr(self, "_endpoint_service_type") or len(self._endpoint_service_type) == 0 or self._endpoint_service_type.isspace():
+            self._endpoint_service_type = self._retrieve_service_type()
+
+        if not self._EXPECTED_SERVICE_TYPE == self._endpoint_service_type:
+            message = self._WRONG_ENDPOINT_MESSAGE.format(self._EXPECTED_SERVICE_TYPE, self._endpoint_service_type)
+            has_endpoint = True
+            if not hasattr(self, "_suggested_endpoint_uri") or len(self._suggested_endpoint_uri) == 0 or self._suggested_endpoint_uri.isspace():
+                self._suggested_endpoint_uri = self._generate_endpoint_suggestion(self._connection_datasource)
+                if not hasattr(self, "_suggested_endpoint_uri") or len(self._suggested_endpoint_uri) == 0 or self._suggested_endpoint_uri.isspace():
+                    has_endpoint = False
+            if has_endpoint:
+                message = "{0}: '{1}'".format(message, self._suggested_endpoint_uri)
+            else:
+                message += "."
+            raise KustoClientError(message)
+
+    def _retrieve_service_type(self):
```

Comment:
Just call the underlying single line of code

Reply:
"Just call the underlying single line of code"
This is "hachana l'mazgan". With this approach, we can add the same feature to the Streaming Ingest Client without duplicating code. I took this approach in the Java SDK as well. The parent class will define the same signature, but the Streaming Client's implementation will be more complex. All the other functions I added to this file (with 1 small exception) can then be moved to the parent class, as the code's identical between the 2 clients.

Reply:
"Since this is a fault flow which needs fixing by the user, I see no point in caching the message creation."
The message isn't cached. The _endpoint_service_type and _suggested_endpoint_uri are cached similar to the .NET SDK so that if the code using the SDK executes many calls and the client with the incorrect endpoint is a non-critical part of that (e.g. the critical flow they're doing involves a correctly-configured client, and so they won't catch the issue with the incorrectly-configured client right away), it won't require a slow ".show version" lookup and suggestion computation every time it *is* used.

I agree with you that it's not a big win since it's an error state they'll have to fix anyway, but I think it is better, and I also preferred consistency with the other SDKs. WDYT?

Reply:
If you want to cache drop the hasattr syntax and create an empty class member in the initialization

Reply:
Resolving, as this is a duplicate of the discussion above about pros/cons of defining each class instance variable in the init vs. keeping variables that are irrelevant to the happy path and only used in the wrong-endpoint-detection feature from dirtying up the rest of the class.